crates/test-environment/src/manifest_template.rs

```diff
@@ -67,7 +67,51 @@ impl EnvTemplate {
         Ok(())
     }
 
+    pub fn substitute_value(
+        &mut self,
+        key: &str,
+        replacement: impl Fn(&str) -> String,
+    ) -> anyhow::Result<()> {
+        replace_template(&mut self.content, |k, v| {
+            if k == key {
+                Ok(Some(replacement(v)))
+            } else {
+                Ok(None)
+            }
+        })
+    }
+
     pub fn contents(&self) -> &str {
         &self.content
     }
 }
+
+/// Replace template variables in a string.
+///
+/// Every time a template is found, the `replacement` function is called with the template key and value.
+pub fn replace_template(
+    content: &mut String,
+    mut replacement: impl FnMut(&str, &str) -> anyhow::Result<Option<String>>,
+) -> Result<(), anyhow::Error> {
+    let regex = TEMPLATE_REGEX.get_or_init(|| regex::Regex::new(r"%\{(.*?)\}").unwrap());
+    'outer: loop {
+        'inner: for captures in regex.captures_iter(content) {
```

Comment:
This might be less confusing if you break out the iterator, e.g.

```suggestion
    'outer: loop {
        let mut captures_iter = regex.captures_iter(content);
        'inner: for captures in captures_iter {
```